File: app/src/main/java/com/kota/Bahamut/Command/BahamutCommandTheSameTitleDown.java
package com.kota.Bahamut.Command;

import androidx.annotation.NonNull;

import com.kota.ASFramework.Thread.ASRunner;
import com.kota.ASFramework.UI.ASToast;
import com.kota.Bahamut.ListPage.TelnetListPage;
import com.kota.Bahamut.ListPage.TelnetListPageBlock;
import com.kota.Telnet.Reference.TelnetKeyboard;
import com.kota.Telnet.TelnetOutputBuilder;

public class BahamutCommandTheSameTitleDown extends TelnetCommand {
    int _article_index = 0;

    public BahamutCommandTheSameTitleDown(int articleIndex) {
        Action = TheSameTitleDown;
        _article_index = articleIndex;
    }

    public void execute(final TelnetListPage aListPage) {
        if (aListPage.getListType()>0) {
            if (_article_index == aListPage.getItemSize()) {
                new ASRunner() {
                    public void run() {
                        ASToast.showShortToast("無下一篇同主題文章");
                        aListPage.onLoadItemFinished();
                    }
                }.runInMainThread();
                setDone(true);
            } else {
                TelnetOutputBuilder.create()
                        .pushString(_article_index + "\n")
                        .pushKey(TelnetKeyboard.DOWN_ARROW).sendToServer();
            }
        } else if (_article_index > 0) {
            TelnetOutputBuilder.create()
                    .pushString(_article_index + "\n]")
                    .sendToServer();
        } else {
            setDone(true);
        }
    }

    public void executeFinished(final TelnetListPage aListPage, TelnetListPageBlock aPageData) {
        if (aPageData.selectedItem.isDeleted || aListPage.isItemBlocked(aPageData.selectedItem)) {
            if (_article_index == aPageData.selectedItemNumber) {
                new ASRunner() {
                    public void run() {
                        ASToast.showShortToast("無下一篇同主題文章");
                        aListPage.onLoadItemFinished();
                    }
                }.runInMainThread();
                setDone(true);
            } else {
                _article_index = aPageData.selectedItemNumber;
                setDone(false);
            }
        } else if (aListPage.isItemLoadingByNumber(aPageData.selectedItemNumber)) {
            new ASRunner() {
                public void run() {
                    ASToast.showShortToast("無下一篇同主題文章");
                    aListPage.onLoadItemFinished();
                }
            }.runInMainThread();
            setDone(true);
        } else if (!aListPage.isEnabled(aPageData.selectedItemNumber - 1)) {
            ASToast.showShortToast("下一篇不可使用");
            aListPage.onLoadItemFinished();
            setDone(true);
        } else {
            aListPage.loadItemAtNumber(aPageData.selectedItemNumber);
            setDone(true);
        }
    }

    @NonNull
    public String toString() {
        return "[TheSameTitleDown][articleIndex=" + _article_index + "]";
    }
}

File: app/src/main/java/com/kota/Bahamut/command/BahamutCommandTheSameTitleDown.kt
package com.kota.Bahamut.command

import com.kota.Bahamut.listPage.TelnetListPage
import com.kota.Bahamut.listPage.TelnetListPageBlock
import com.kota.asFramework.thread.ASCoroutine
import com.kota.asFramework.ui.ASToast.showShortToast
import com.kota.telnet.TelnetOutputBuilder.Companion.create
import com.kota.telnet.reference.TelnetKeyboard

class BahamutCommandTheSameTitleDown(fromArticleIndex: Int) : TelnetCommand() {
    var articleIndex: Int = 0

    init {
        action = BahamutCommandDef.Companion.THE_SAME_TITLE_DOWN
        articleIndex = fromArticleIndex
    }

    override fun execute(telnetListPage: TelnetListPage) {
        if (telnetListPage.listType > 0) {
            if (articleIndex == telnetListPage.listCount) {
                ASCoroutine.ensureMainThread {
                    showShortToast("無下一篇同主題文章")
                    telnetListPage.onLoadItemFinished()
                }
                isDone = true
            } else {
                create()
                    .pushString(articleIndex.toString() + "\n")
                    .pushKey(TelnetKeyboard.DOWN_ARROW).sendToServer()
            }
        } else if (articleIndex > 0) {
            create()
                .pushString("$articleIndex\n]")
                .sendToServer()
        } else {
            isDone = true
        }
    }

    override fun executeFinished(telnetListPage: TelnetListPage, telnetListPageBlock: TelnetListPageBlock?) {
        if (telnetListPageBlock?.selectedItem?.isDeleted == true || telnetListPage.isItemBlocked(
                telnetListPageBlock?.selectedItem
            )) {
            if (articleIndex == telnetListPageBlock?.selectedItemNumber) {
                ASCoroutine.ensureMainThread {
                    showShortToast("無下一篇同主題文章")
                    telnetListPage.onLoadItemFinished()
                }
                isDone = true
            } else {
                articleIndex = telnetListPageBlock?.selectedItemNumber!!
                isDone = false
            }
        } else if (telnetListPage.isItemLoadingByNumber(telnetListPageBlock?.selectedItemNumber!! )) {
            ASCoroutine.ensureMainThread {
                showShortToast("無下一篇同主題文章")
                telnetListPage.onLoadItemFinished()
            }
            isDone = true
        } else if (!telnetListPage.isEnabled(telnetListPageBlock.selectedItemNumber - 1)) {
            showShortToast("下一篇不可使用")
            telnetListPage.onLoadItemFinished()
            isDone = true
        } else {
            telnetListPage.loadItemAtNumber(telnetListPageBlock.selectedItemNumber)
            isDone = true
        }
    }

    override fun toString(): String {
        return "[TheSameTitleDown][articleIndex=$articleIndex]"
    }
}
